### ISSUE TITLE
Expose detected compiler executable name?

### ISSUE THREAD

**Issue by torokati44** [mesonbuild/meson#3215]
In the framework I'm porting from autoconf+make to meson, the `configure` script generates a `Makefile.inc` file. This is used both for building the framework itself, and is then provided to the users of the framework. This way their applications don't need a configuration step, and are guaranteed to share the same configuration as the framework itself - which is helpful since they are usually linked together in the end. I don't want to make using meson mandatory for users, even if the framework was built with it.

To remain compatible, I'm trying to generate this `Makefile.inc` as a configuration file from meson. This must also contain the names of the C and C++ compilers to use. Basically I would have to set a configuration variable to the same compiler name that is emitted into the `build.ninja` file, for example `cl.exe`, `ccache clang++` or whatever came from the `CC` and `CXX` environment variables.

I couldn't find a way to access this information in meson. Is this possible right now? If not, I'd be happy to try implementing this. I'm thinking of adding a method named something like `exe_name()` to `compiler` objects. Is this something you would accept?

**Comment by torokati44** [comment]
And this is likely just one of the several things I would need. The linker command, and many more would also be necessary. Are you okay with extending meson in this direction, so it can more-or-less act as a `configure` script?

**Comment by jpakkane** [comment]
The common solution for this is providing pkg-config files instead. What is the specific reason that you must enforce the same C compiler for building the framework as well as code that uses it? It seems very restrictive.

Adding more data is something we want but on the other hand we try to avoid "string typing" as much as possible. That is, we'd prefer if you could use the compiler object directly in commands and have it do the right thing rather than people manually building strings to do stuff and then getting it wrong (this is a repeating problem case, for some reason _everyone_ really wants to create strings manually even when it makes no sense).

**Comment by torokati44** [comment]
It's mostly convenience and ease of use. Technically they don't _have_ to use _the same_ compiler, but the way we think about it is that the framework provides the compiler as well as its setup to the users. This way they don't have to configure anything, just write a bunch of `.h` and `.cc` files, and the rest is taken care of, regardless of the platform, or whether there is cross-compilation, or if a custom compiler [wrapper] is used.

Can `pkg-config` also define the compiler to be used? I thought it can only provide arguments for it, which are necessary to use a library.

I agree that string-oriented programming is bad, however, at the moment the output would be a text file, included by make (when compiling the user applications), so there really isn't much choice other than writing strings into it.

In the long run, it might be better to have the user applications use meson for building as well, but at the moment our tooling interoperates fairly tightly with the way the Makefiles are set up for the user applications.

**Comment by nirbheek** [comment]
> Can pkg-config also define the compiler to be used? I thought it can only provide arguments for it, which are necessary to use a library.

No, but you *can* define arbitrary variables in it, and people do use that to store paths to tools that users of their project will need. It sounds like this would work quite well for you since it's the same use-case.

For example: https://gitlab.gnome.org/GNOME/gobject-introspection/blob/master/gobject-introspection-1.0.pc.in#L9

**Comment by torokati44** [comment]
I see. So what you're suggesting is that the `pkgconfig` module should be extended so it is able to accept a `compiler` object, and record any interesting information about it (like the command used to run it) in a `.pc` file as custom values, with some predefined keys?

**Comment by nirbheek** [comment]
The pkgconfig module can already define arbitrary variables with the `variables:` kwarg. You should be able to use that.

**Comment by torokati44** [comment]
Yes, but then we are back at string juggling, and I still don't have access to the command used to run the compiler. The best I could do is guess based on the platform and the compiler id.

And this way there is no need for a `.pc` file, since our current build system works with a `Makefile.inc`. It would be just more work to set the `CC` and `CXX` variables in it using `pkg-config`, compared to generating it from meson as a configuration file.

**Comment by nirbheek** [comment]
> Yes, but then we are back at string juggling, and I still don't have access to the command used to run the compiler.

Ah yes, sorry. I forgot that the format of that kwarg is `VAR=value`. I guess we'd need some way of getting that value in there without using strings.

**Comment by torokati44** [comment]
Like... using some sort of placeholder, for example `CompilerCommand=${compiler_command}`? 😕 Or what is the idea? 🤔

**Comment by nirbheek** [comment]
We've been talking about an associative-array type for kwargs that need that sort of `key: value` model. I think that would be the first step, but it depends on what @jpakkane thinks.

**Comment by torokati44** [comment]
Has there been any discussion about this feature recently? I'm still interested in it (for the long term).

**Comment by nirbheek** [comment]
A dict type was merged in this cycle, so the next step is accepting dicts for `variables:` in the pkgconfig module (which is easy), and then adding support for accepting the `CompilerHolder` type in the same module (which is also easy). This was previously done for `run_command()`, so the same code can be used here. See, `mesonbuild/interpreter.py`:`run_command_impl()` if you want to have a go at it.

**Comment by nirbheek** [comment]
But, actually, I think for your case (to generate `Makefile.inc` from `Makefile.inc.in`), a better mechanism would be for `configuration_data()` objects to understand `CompilerHolder` so that it converts the object to a string while writing it out with `configure_file()`.

**Comment by eli-schwartz** [comment]
Am I missing something here? It seems to me that this was always possible. `compiler.cmd_array()` has existed since commit 572587f5cef0741cf68d5842d503ddb764e3d47b which was years before this ticket was filed.

```meson
project('foo', 'c')

cc = meson.get_compiler('c')

cdata = configuration_data()
cdata.set_quoted('CC', ' '.join(cc.cmd_array()))

configure_file(output: 'config.h', configuration: cdata)
```

Produces: 
```h
#define CC "ccache /home/eschwartz/git/meson/symlinked cc/gcc"
```
Yes, this includes the obvious breakage...

**Comment by xclaesse** [comment]
> Yes, this includes the obvious breakage...

We could maybe add `str.shell_join()` and `str.shell_split()` for this type of usage?

**Comment by eli-schwartz** [comment]
> We could maybe add `str.shell_join()` and `str.shell_split()` for this type of usage?

Also:

There's no real way to predict which of various different quoting styles are valid in different output formats especially when not using a header/json template but using `@REPLACEMENT@` tokens. So I'm not even sure this would help. A better notion would be to record each element separately and have the project code (e.g. C++) assemble it as desired.

`output_format: 'json'` is new, and could in theory support structured values, but those aren't accepted by configuration_data anyway...
